File: app/llmApi/request.js
const axios = require('axios');
const fs = require('fs');
const Handlebars = require('handlebars');
require('dotenv').config();

const OPENAI_TOKEN = process.env.OPENAI_TOKEN;
const promptTextfile = './prompt.txt';

/**
 * OpenAI APIにリクエストを送信する
 * @param {string} protoFileContent - .protoファイルの内容
 * @param {string} fileChangesContent - 変更ファイルの内容
 * @param {string} sourceCodeContent - ソースコードの内容
 */
async function requestOpenAI(protoFileContent, fileChangesContent, sourceCodeContent) {
    const endpoint = '/v1/chat/completions';
    const url = `https://api.openai.com${endpoint}`;

    let fileContent = fs.readFileSync(promptTextfile, 'utf-8');

    // コンテキストとして渡すデータ
    let context = {
        protoFile: protoFileContent,
        fileChanges: fileChangesContent,
        sourceCode: sourceCodeContent
    };

    // Handlebarsテンプレートをコンパイルしてプロンプト生成
    let template = Handlebars.compile(fileContent);
    let prompt = template(context);

    const payload = {
        'model': 'gpt-4o',
        'messages': [
            {
                "role": "system",
                "content": "Fix or improve issues in the program code related to gRPC, please refer to the proto file and make the code fixes."
            },
            {
                'role': 'user',
                'content': `${prompt}`
            }
        ]
    };

    const config = {
        headers: {
            'Content-Type': 'application/json',
            'Authorization': `Bearer ${OPENAI_TOKEN}`
        }
    };

    try {
        const response = await axios.post(url, payload, config);
        console.log(response.data);
        console.log(response.data.choices[0].message.content);
    } catch (error) {
        console.error(error.response ? error.response.data : error.message);
    }
}

// モジュールとしてエクスポート
module.exports = {
    requestOpenAI
};

File: app/llmApi/python/request.py
#Open AI APIを利用したリクエスト

import os
import requests
from jinja2 import Template
from dotenv import load_dotenv

# .envファイルから環境変数をロード
load_dotenv()

OPENAI_TOKEN = os.getenv('OPENAI_TOKEN')
PROMPT_TEXT_FILE = '/app/app/llmApi/prompt.txt'

def request_openai(proto_file_content, file_changes_content, source_code_content):
    """
    OpenAI APIにリクエストを送信する

    :param proto_file_content: .protoファイルの内容
    :param file_changes_content: 変更ファイルの内容
    :param source_code_content: ソースコードの内容
    :return: OpenAI APIのレスポンス内容（成功した場合）または0（失敗した場合）
    """
    endpoint = '/v1/chat/completions'
    url = f'https://api.openai.com{endpoint}'

    try:
        # プロンプトテキストファイルの読み込み
        with open(PROMPT_TEXT_FILE, 'r', encoding='utf-8') as file:
            file_content = file.read()

        # Jinja2テンプレートのコンパイルとコンテキストの適用
        context = {
            'protoFile': proto_file_content,
            'fileChanges': file_changes_content[0]['diff'],
            'sourceCode': source_code_content
        }

        template = Template(file_content)
        prompt = template.render(context)

        payload = {
            'model': 'gpt-4',
            'messages': [
                {
                    'role': 'system',
                    'content': 'Fix or improve issues in the program code related to gRPC, please refer to the proto file and make the code fixes.'
                },
                {
                    'role': 'user',
                    'content': prompt
                }
            ]
        }

        headers = {
            'Content-Type': 'application/json',
            'Authorization': f'Bearer {OPENAI_TOKEN}'
        }

        # OpenAI APIへPOSTリクエストを送信
        response = requests.post(url, json=payload, headers=headers)
        response.raise_for_status()

        # APIレスポンスを取得して出力
        data = response.json()
        print(data)
        print(data['choices'][0]['message']['content'])

        return data['choices'][0]['message']['content']

    except requests.exceptions.RequestException as e:
        print(f"Error during the request: {e}")
        return 0

    except Exception as e:
        print(f"Unexpected error: {e}")
        return 0


# モジュールとしてエクスポートできるようにする
if __name__ == '__main__':
    # テスト用に関数を呼び出し
    proto_content = "syntax = 'proto3';"
    file_changes = "some file changes"
    source_code = "def some_function(): pass"

    result = request_openai(proto_content, file_changes, source_code)
    print("OpenAI Response:", result)
